### ISSUE TITLE
Loss scalings

### ISSUE THREAD

**Issue by casperkaae** [jakobnissen/vamb#1]
in  https://github.com/jakobnissen/vamb/blob/master/encode.py#L174-L180
```    
def calc_loss(self, depths_in, depths_out, tnf_in, tnf_out, mu, logsigma):
        ce = - _torch.mean((depths_out.log() * depths_in))
        mse = _torch.mean((tnf_out - tnf_in).pow(2))
        kld = -0.5 * _torch.mean(1 + logsigma - mu.pow(2) - logsigma.exp())
        loss = self.cefactor * ce + self.msefactor * mse + self.kldfactor * kld

        return loss, ce, mse, kld
```
The cross-entropy should be  `ce = - _torch.mean((depths_out.log() * depths_in))`

We can then change https://github.com/jakobnissen/vamb/blob/master/encode.py#L381
```
cefactor = nsamples * (1 - mseratio) / _log(nsamples)
```
to 
```
cefactor = (1 - mseratio) / _log(nsamples)
```

Why is the _log(nsamples) scaling term there? does it help otherwise i think it introduces a weird dependency on the number of samples?

**Comment by jakobnissen** [comment]
@casperkaae I'm not sure I understand what you mean. You propose *no change* to the calculation of `ce`? Or did you intend to recommend replacing `_torch.mean` with `torch.sum`?

Assuming you did so:
Yes, we could do that. Your proposed change to `cefactor` would make the resulting code equivalent to the existing code, causing no change in behaviour. However, we would then have to explain why we use `mean` on the MSE and KLD losses and `sum` on the CE loss. Obviously, we could *also* use `sum` on *those* losses, but that would be equivalent to just multiplying the entire loss with `nsamples`. This would cause no change in behaviour, except that the total loss would now be dependent on the number of samples, which can vary wildly between datasets. This would have to be combatted by decreasing the learning rate or similar. The overall result would be that the behaviour would be unchanged, and the code would be more complicated.

From where do we get the loss scalings in the first place? Our reason is actually kind of hand-wavey, and I would appreciate any suggestions to do differently. Here's the reasoning:
We want to scale CE, MSE and KLD such that e.g. 90% of the loss comes from CE rather than MSE. They obviously have different numerical scales in the first place, so if we want `CE == 9 * MSE`, we can just calculate CE and MSE, and then scale them to fit. But what *is* the value of MSE and CE? They change during training (that's the point of training!). What we do is to *calculate the expected MSE and CE of an untrained network*. The expected MSE is (N(0,1)-N(0,1))^2 = 1, and if we assume the network returns [1/nsamples, 1/nsamples ... ] for the abundances, the expected CE is -sum(ln(1/nsamples) * abundance)/nsamples = sum(abundance) * -sum(-ln(nsamples)) / nsamples = ln(nsamples)/nsamples.

Therefore, if we want it to be the case that the expected MSE is `mseratio` and the expected CE is `1 - mseratio`, we scale `MSE = mseratio * MSE` and `CE = nsamples * (1 - mseratio) / ln(nsamples)`.

**Comment by casperkaae** [comment]
My intention was to
1) calculate the individual terms i the ELBO "correctly". this also involves summing in KL and MSE. Actually the MSE also needs a anothor scaling factor if we wish to interpret it is a gaussian likelihood as assumed in the ELBO
2) introduce a single weighing factor for each term i the loss that captures the scalings that are in different places in code and manuscript now

**Comment by jakobnissen** [comment]
Regarding restructuring the code, you are right. It will make it easier to understand conceptually.
About the new scaling factor for MSE, I'm interested in hearing more. What scaling factor?

**Comment by casperkaae** [comment]
MSE is essentially the same as assuming a gaussian likelihood model i.e. we calculate the likelihood of the data under the gaussian distribution parameterized by our model. The gaussian PDF is

![image](https://user-images.githubusercontent.com/3503745/45812599-82e40080-bcd0-11e8-8ba2-78779a9eaa7c.png)

We are maximizing the log of that expression .e.g

![image](https://user-images.githubusercontent.com/3503745/45812786-11588200-bcd1-11e8-9ed4-7512c41835c9.png)

In our case we only predict the mean from the the model and (implicitly) assume that the variance is always 1. In our case the scaling on the squared error terms is then actually 1 - we can safely ignore the constant in front since it only depends on the variance - in principle we should add it to the loss when we plot it but it does not matter for the optimization.
